tracer/src/Datadog.Trace/DataStreamsMonitoring/Aggregation/DataStreamsMessagePackFormatter.cs

```diff
@@ -84,26 +88,36 @@ public int Serialize(Stream stream, long bucketDurationNs, List<SerializableStat
             bytesWritten += MessagePackBinary.WriteStringBytes(stream, _tracerVersionValueBytes);
 
             bytesWritten += MessagePackBinary.WriteStringBytes(stream, _statsBytes);
-            bytesWritten += MessagePackBinary.WriteArrayHeader(stream, statsBuckets.Count);
+            bytesWritten += MessagePackBinary.WriteArrayHeader(stream, statsBuckets.Count + backlogsBuckets.Count);
 
-            foreach (var statsBucket in statsBuckets)
+            foreach (var backlogBucket in backlogsBuckets)
             {
-                // 3 entries per StatsBucket:
-                // https://github.com/DataDog/data-streams-go/blob/6772b163707c0a8ecc8c9a3b28e0dab7e0cf58d4/datastreams/payload.go#L27
-                bytesWritten += MessagePackBinary.WriteMapHeader(stream, 3);
+                bytesWritten += WriteBucketsHeader(stream, backlogBucket.BucketStartTimeNs, bucketDurationNs, 0, backlogBucket.Bucket.Values.Count);
 
-                bytesWritten += MessagePackBinary.WriteStringBytes(stream, _startBytes);
-                bytesWritten += MessagePackBinary.WriteInt64(stream, statsBucket.BucketStartTimeNs);
+                foreach (var point in backlogBucket.Bucket.Values)
+                {
+                    bytesWritten += MessagePackBinary.WriteMapHeader(stream, 2);
 
-                bytesWritten += MessagePackBinary.WriteStringBytes(stream, _durationBytes);
-                bytesWritten += MessagePackBinary.WriteInt64(stream, bucketDurationNs);
+                    bytesWritten += MessagePackBinary.WriteStringBytes(stream, _backlogValueBytes);
+                    bytesWritten += MessagePackBinary.WriteInt64(stream, point.Value);
 
-                bytesWritten += MessagePackBinary.WriteStringBytes(stream, _statsBytes);
-                bytesWritten += MessagePackBinary.WriteArrayHeader(stream, statsBucket.Bucket.Values.Count);
+                    var tags = point.Tags.Split(',');
```

Comment:
Calling `Split` is rarely a good idea, as it creates a lot of garbage, but I'm not sure if it's possible to write a substring with our current `MessagePackBinary` 🤔 Worth looking into...

Reply:
I used string here to avoid dealing with an array of tags for each `Backlog` object. I would also need to hash that list to perform the aggregation. Using `Split` during data flush seems like a lightweight alternative, especially for a high throughput service. When you say garbage, do you mean extra arrays allocations? I guess converting to a byte array and using WriteStringBytes with offsets will work better here.

Reply:
> Using Split during data flush seems like a lightweight alternative, especially for a high throughput service. When you say garbage, do you mean extra arrays allocations?

`Split` is fine in low-throughput scenarios, but we avoid it wherever possible because it causes
- an allocation for the array
- an allocation for each individual `string` in the array

> I guess converting to a byte array and using WriteStringBytes with offsets will work better here.

I don't think that's necessarily a good idea, as it's a bit different (one string vs many strings). Unfortunately, it looks lie there isn't good support for partial strings or `Span` in our current vendored `MessagePackBinary` version, so I'm not sure you can do much better 🙁 

About the best you can do is avoid the array allocation using IndexOf(',') and `SubString()`, but that doesn't save you a lot I guess, so maybe isn't worth it.

Reply:
Due the fact we are not going to update the messagepack library because that would require a major refactor and extra work to support older framework. We can introduce some extensions in the current version, it doesn't seem hard to create an overload for https://github.com/DataDog/dd-trace-dotnet/blob/master/tracer/src/Datadog.Trace/Vendors/MessagePack/MessagePackBinary.cs#L1894 accepting an startIndex and stringlength to write substrings.

That said, I wouldn't do that in this PR.